Core/GDCore/IDE/Events/UsedExtensionsFinder.cpp

```diff
@@ -0,0 +1,143 @@
+#include "UsedExtensionsFinder.h"
+
+#include "GDCore/Events/Instruction.h"
+#include "GDCore/Events/Parsers/ExpressionParser2.h"
+#include "GDCore/Extensions/Metadata/MetadataProvider.h"
+#include "GDCore/Extensions/PlatformExtension.h"
+#include "GDCore/IDE/WholeProjectRefactorer.h"
+#include "GDCore/Project/BehaviorContent.h"
+#include "GDCore/Project/Object.h"
+#include "GDCore/Project/Project.h"
+
+namespace gd {
+
+std::set<gd::String> UsedExtensionsFinder::ScanProject(gd::Project& project) {
+  UsedExtensionsFinder worker(project);
+  gd::WholeProjectRefactorer::ExposeProjectObjects(project, worker);
+  gd::WholeProjectRefactorer::ExposeProjectEvents(project, worker);
+  return worker.usedExtensions;
+};
+
+// Objects scanner
+
+void UsedExtensionsFinder::DoVisitObject(gd::Object& object) {
+  usedExtensions.insert(gd::MetadataProvider::GetExtensionAndObjectMetadata(
+                            project.GetCurrentPlatform(), object.GetType())
+                            .GetExtension()
+                            .GetName());
+};
+
+// Behaviors scanner
+
+void UsedExtensionsFinder::DoVisitBehavior(gd::BehaviorContent& behavior) {
+  usedExtensions.insert(
+      gd::MetadataProvider::GetExtensionAndBehaviorMetadata(
+          project.GetCurrentPlatform(), behavior.GetTypeName())
+          .GetExtension()
+          .GetName());
+};
+
+// Instructions scanner
+
+bool UsedExtensionsFinder::DoVisitInstruction(gd::Instruction& instruction,
+                                              bool isCondition) {
+  auto metadata =
+      isCondition ? gd::MetadataProvider::GetExtensionAndConditionMetadata(
+                        project.GetCurrentPlatform(), instruction.GetType())
+                  : gd::MetadataProvider::GetExtensionAndActionMetadata(
+                        project.GetCurrentPlatform(), instruction.GetType());
+  usedExtensions.insert(metadata.GetExtension().GetName());
+
+  size_t i = 0;
+  for (auto expression : instruction.GetParameters()) {
+    const gd::String& parameterType =
+        metadata.GetMetadata().GetParameter(i++).GetType();
```

Comment:
Nitpicking: let's avoid this clever i++ because it's easy to miss (when copy/pasting), rely on knowledge of somewhat obscure C++ corners (postfix ++ returns the original value - but that's a contract that could be not honored by some reimplementations), and so can lead to hard to debug code ("I've duplicated this line which job is to get the parameter type and now for some reason I'm missing some parameters").

A i++; on the next line will make extra explicit that we're iterating using two iterators (`expression` from the for each range, and `i`), which is an unusual construct anyway.